probe/host/host.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2022, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package host
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/megaease/easeprobe/global"
+	"github.com/megaease/easeprobe/probe"
+	"github.com/megaease/easeprobe/probe/ssh"
+	log "github.com/sirupsen/logrus"
+)
+
+// Default threshold
+const (
+	DefaultCPUThreshold  = 0.8
+	DefaultMemThreshold  = 0.8
+	DefaultDiskThreshold = 0.95
+)
+
+// Threshold is the threshold of a probe
+type Threshold struct {
+	CPU  float64 `yaml:"cpu"`
+	Mem  float64 `yaml:"mem"`
+	Disk float64 `yaml:"disk"`
+}
+
+func (t *Threshold) String() string {
+	return fmt.Sprintf("CPU: %.2f, Mem: %.2f, Disk: %.2f", t.CPU, t.Mem, t.Disk)
+}
+
+// Server is the server of a host probe
+type Server struct {
+	ssh.Server `yaml:",inline"`
+	Threshold  Threshold `yaml:"threshold"`
+}
+
+// Host is the host probe configuration
+type Host struct {
+	Bastion *ssh.BastionMapType `yaml:"bastion"`
+	Servers []Server            `yaml:"servers"`
+}
+
+// BastionMap is a map of bastion
+var BastionMap ssh.BastionMapType
+
+// Config is the host probe configuration
+func (s *Server) Config(gConf global.ProbeSettings) error {
+	kind := "host"
+	tag := "server"
+	name := s.ProbeName
+
+	// The following commands are:
+	// 1. retrive the hostname: 	`hostname``
+	// 2. retrive the os:  			`awk -F= '/^NAME/{print $2}' /etc/os-release | tr -d '\"'`
+	// 3. retrive the memory usage:	`free -m | awk 'NR==2{printf "%s %s %.2f\n", $3,$2,$3*100/$2 }'`
+	// 4. retrive the cpu core:		`grep -c ^processor /proc/cpuinfo;`
+	// 5. retrive the cpu usage:	`top -b -n 1 | grep Cpu | awk -F ":" '{print $2}'`
+	// 6. retrive the disk usage	`df -h | awk '$NF=="/"{printf "%d %d %s\n", $3,$2,$5}'`
+	s.Command = `hostname;
+	awk -F= '/^NAME/{print $2}' /etc/os-release | tr -d '\"';
+	free -m | awk 'NR==2{printf "%s %s %.2f\n", $3,$2,$3*100/$2 }';
+	grep -c ^processor /proc/cpuinfo;
+	top -b -n 1 | grep Cpu | awk -F ":" '{print $2}';
```

Comment:
`Top` command is required. Why didn't calculate the CPU usage from `/proc/stat` [1] with the pure script?

[1]https://stackoverflow.com/questions/11356330/how-to-get-cpu-usage

Reply:
Using `/proc/stat` to calculate the CPU usage is so complicated, and there are many different methods to do that. And I can see different calculations have different results. I think it might be better to align with the actual operation - when people check the CPU usage, everyone uses `top` instead of `/proc/stat`,  I think using `top` can have the same picture with the operation, and avoid the confusion. That's why I decide to use `top` instead of `/proc/stat`. 

[1] https://stackoverflow.com/questions/9229333/how-to-get-overall-cpu-usage-e-g-57-on-linux (there are many calculations in the answers replies)
[2] https://supportcenter.checkpoint.com/supportcenter/portal?eventSubmit_doGoviewsolutiondetails=&solutionid=sk65143
[3] Another good article to compare the `top` and `/proc/stat` https://medium.com/@yogita088/how-to-calculate-cpu-usage-proc-stat-vs-top-e74f99f02d08

Reply:
Leveraging the top command brings a consistent result, I totally agree with you, but it needs an extra component, this is a trade-off. So we should clarify that the probe based on infrastructure metrics needs to ensure the machine has installed the top command.

Reply:
I will clarify that in the document.